### ISSUE TITLE
Intermittent duplicate Modules in XML report

### ISSUE THREAD

**Issue by squid-box** [OpenCover/opencover#908]
### My Framework

- [ ] .NET 2
- [ ] .NET 3.5
- [ ] .NET 4
- [ ] .NET 4.5
- [ ] .NET 4.6
- [ ] .NET 4.6.1
- [ ] .NET 4.6.2
- [ ] .NET 4.7
- [ ] .NET 4.7.1
- [X] .NET 4.7.2
- [ ] .NET Core 1.0.0
- [ ] .NET Core 2.0.0
- [ ] Something else

### My Environment

- [ ] Windows 7 or below (not truly supported due to EOL)
- [ ] Windows 8
- [ ] Windows 8.1
- [X] Windows 10
- [ ] Windows 10 IoT Core
- [ ] Windows Server 2012
- [ ] Windows Server 2012 R2
- [ ] Windows Server 2016

### I have already...

- [X] repeated the problem using the latest stable release of OpenCover.
- [X] reviewed the [usage guide](https://github.com/opencover/opencover/wiki/usage) and [usage document](https://github.com/opencover/opencover/blob/master/main/OpenCover.Documentation/Usage.pdf).
- [X] have looked at the opencover output xml file in an attempt to resolve the issue.
- [X] reviewed the [current issues](https://github.com/opencover/opencover/issues) to check that the issue isn't already known.

### My issue is related to (check only those which apply):

- [ ] no coverage being recorded
- [ ] 32 or 64 bit support
- [ ] feature request

### Expected Behavior

No duplicate modules in the XML output, if `-mergebyhash` is supplied.

### Actual Behavior

Intermittently (something like every 3rd - 5th run) the XML report will include duplicate instances of certain modules (seems to always be for the same file). There's 2-3 instances of the identical module, where only one of them has non-zero coverage data. The biggest issue for my use-case is that this affects the summary values.

The module duplicates are, to my eyes, identical. They have the same hash, name/location (even in casing), timestamp, and file list (same file names, different uid's).

When we noticed the issue we were not using `-mergebyhash`, but even with it enabled it occurs.

### Steps to reproduce the problem:

- [X] reviewed the [usage guide](https://github.com/opencover/opencover/wiki/usage) and [usage document](https://github.com/opencover/opencover/blob/master/main/OpenCover.Documentation/Usage.pdf) again because we know you probably didn't do it the first time.
- [ ] sample code or link to repository/gist is available (support is much faster this way) 

Unfortunately I've encountered this working on proprietary code, so I can't (easily) provide a sample reproducing this. The way we're using OpenCover is:

We have a solution with several .NET 4.7.2 projects inside, including multiple NUnit test projects. This solution is being built & tested with [Cake|https://cakebuild.net/], which also runs OpenCover (though I'm seeing the same behavior running OpenCover entirely standalone on the same code).

The OpenCover call is (censored):
```
"<Somepath>/Tools/OpenCover.4.7.922/tools/OpenCover.Console.exe" -target:"<Somepath>/Tools/NUnit.ConsoleRunner.3.9.0/tools/nunit3-console.exe" -targetargs:"\"One.Tests.dll\" \"Two.Tests.dll\" \"Three.Tests.dll\" --timeout=60000 \"--work=<OtherPath>/Stage/NUnit\" \"--result=<OtherPath>/NUnit/NUnitResults.xml\" \"--result=<OtherPath>/Stage/NUnit/NUnitResults.html;transform=<AnotherPath>/XSLT/NUnitReport.xslt\" --labels=All --noheader --x86" -filter:"-[*.Tests]* -[*]XamlGeneratedNamespace* +[Product.*]* +[OtherPart*]*" -register:user -output:"<OtherPath>/NUnit/OpenCover.xml" -mergebyhash -hideskipped:All
```

The versions we're using are:
* OpenCover: 4.7.922 (through NuGet)
* NUnit.ConsoleRunner 3.9.0

**Comment by sawilde** [comment]
If you can give me a way to repeat it I can have a go otherwise you are on your own - the source code is available to everyone

**Comment by squid-box** [comment]
I'll take this as confirmation that I've at least understood the usage of `-mergebyhash` :)

I spent some time today trying to create a trivial example showing the behavior, with no luck. I'll try debugging OpenCover and see if I can find more info on what's happening - any suggestions on possible causes are welcome.

**Comment by carlos-quintero** [comment]
I have also experienced this problem and after several hours I was able to isolate, debug and find the root cause. It happens when you have two threads that process the same module with the same hash. In file https://github.com/OpenCover/opencover/blob/master/main/OpenCover.Framework/Persistance/BasePersistance.cs, the line 81:

`CoverageSession.Modules = list.ToArray();`

is not protected by any of the two locks, so it happens that when thread #1 has not reached yet that line, thread #2, which was able to adquire the lock(s), executes this line and finds no module yet with the same hash:

`var existingModule = modules.FirstOrDefault(x => x.ModuleHash == module.ModuleHash);`

Is there any chance for a quick fix for this bug?
